internal/gatewayapi/listeners.go

```diff
@@ -18,144 +18,174 @@ import (
+					result.InvalidListenerConditions[listener.Name] = metav1.Condition{
+						Type:    string(gatewayapi_v1beta1.ListenerConditionConflicted),
+						Status:  metav1.ConditionTrue,
+						Reason:  string(gatewayapi_v1beta1.ListenerReasonProtocolConflict),
+						Message: "All Listener protocols for a given port must be compatible",
+					}
+					conflicted = true
 				}
 			}
-		case gatewayapi_v1beta1.HTTPSProtocolType, gatewayapi_v1beta1.TLSProtocolType:
-			switch {
-			case int(listener.Port) != result.SecurePort:
-				result.InvalidListenerConditions[listener.Name] = metav1.Condition{
-					Type:    string(gatewayapi_v1beta1.ListenerConditionAccepted),
-					Status:  metav1.ConditionFalse,
-					Reason:  string(gatewayapi_v1beta1.ListenerReasonPortUnavailable),
-					Message: "Only one HTTPS/TLS port is supported",
-				}
-			case secureHostnames[hostname] > 1:
+			if conflicted {
+				break
+			}
+
+			// Hostname conflict
+			if ref.Val(listener.Hostname, "") == ref.Val(otherListener.Hostname, "") {
 				result.InvalidListenerConditions[listener.Name] = metav1.Condition{
 					Type:    string(gatewayapi_v1beta1.ListenerConditionConflicted),
 					Status:  metav1.ConditionTrue,
 					Reason:  string(gatewayapi_v1beta1.ListenerReasonHostnameConflict),
-					Message: "Hostname must be unique among HTTPS/TLS listeners",
+					Message: "All Listener hostnames for a given port must be unique",
 				}
+				conflicted = true
 			}
-		default:
-			result.InvalidListenerConditions[listener.Name] = metav1.Condition{
-				Type:    string(gatewayapi_v1beta1.ListenerConditionAccepted),
-				Status:  metav1.ConditionFalse,
-				Reason:  string(gatewayapi_v1beta1.ListenerReasonUnsupportedProtocol),
-				Message: fmt.Sprintf("Listener protocol %q is unsupported, must be one of HTTP, HTTPS or TLS", listener.Protocol),
+
+			if conflicted {
+				break
 			}
 		}
-	}
 
-	return result
-}
+		if conflicted {
+			continue
+		}
+
+		// Add an entry in the Listener name map.
+		var prefix string
+		if listener.Protocol == gatewayapi_v1beta1.HTTPProtocolType {
+			prefix = "http"
+		} else {
+			prefix = "https"
+		}
+		envoyListenerName := fmt.Sprintf("%s-%d", prefix, listener.Port)
+
+		result.ListenerNames[string(listener.Name)] = envoyListenerName
 
-// HostnameDeref returns the hostname as a string if it's not nil,
-// or an empty string otherwise.
-func HostnameDeref(hostname *gatewayapi_v1beta1.Hostname) string {
-	if hostname == nil {
-		return ""
+		// Add the port to the list if it hasn't been added already.
+		found := false
+		for _, port := range result.Ports {
+			if port.Name == envoyListenerName {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			// Map privileged ports (1-1023) to the range 64513-65535 within
+			// the container. All other ports can be used as-is inside the
+			// container.
+			containerPort := listener.Port
+			if containerPort < 1024 {
+				containerPort += 64512
+			}
```

Comment:
I went with the original port mapping scheme from the design doc here, but could still change this to something else.

Reply:
@izturn @wilsonwu @Jean-Daniel same here - let me know if you have any input on the port mapping scheme being proposed here

Reply:
The other option I'm considering is adding 8000 to the Listener's port to determine the port Envoy should listen on (80 -> 8080, 443 -> 8443, etc), wrapping back around to the beginning of the port range for very high ports and skipping ports 1-1023 (since Envoy can't bind to them by default). The nice thing about this is that the standard ports continue to map to the same container ports, which should help avoid some upgrade issues for the normal case.

Reply:
I've now switched to [the scheme described above](https://github.com/projectcontour/contour/pull/5160#discussion_r1191702512) in https://github.com/projectcontour/contour/pull/5160/commits/2d65eb99eb3df2702f06511e87e13323c508037e, as it's more backwards-compatible and results in less downtime during upgrades when using the standard ports.

Reply:
I find it more understandable and cleaner, as standard use cases will have well known values and the limitation (58559-59581 and 57536-58558 are overlapping) would probably almost never have an impact on real use cases.